File: dotnet/examples/ServerConfiguration/Metrics/TopicMetricCollector/ListTopicMetricCollectors.cs
/**
 * Copyright © 2023 - 2024 Diffusion Data Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System;
using System.Threading;
using System.Threading.Tasks;
using static System.Console;
using System.Linq;
using PushTechnology.ClientInterface.Client.Factories;
using PushTechnology.ClientInterface.Client.Features.Metrics;
using static PushTechnology.ClientInterface.Examples.Program;

namespace PushTechnology.ClientInterface.Examples.ServerConfiguration.Metrics.TopicMetricCollector
{
    public sealed class ListTopicMetricCollectors : Example
    {
        public override async Task Run(CancellationToken cancellationToken, string[] args)
        {
            string serverUrl = args[0];

            var session = Diffusion.Sessions
                .Principal("admin")
                .Credentials(Diffusion.Credentials.Password("password"))
                .Open(serverUrl);

            ITopicMetricCollector collector = null;
            string topicSelector = "?my/topic//";

            var builder = Diffusion.NewTopicMetricCollectorBuilder();
            builder = (ITopicMetricCollectorBuilder)builder.ExportsToPrometheus(false);
            builder = builder.GroupByTopicType(true);
            builder = builder.GroupByTopicView(true);
            builder = builder.GroupByPathPrefixParts(15);
            builder = (ITopicMetricCollectorBuilder)builder.MaximumGroups(10);
            collector = builder.Create("Topic Metric Collector 1", topicSelector);

            await session.Metrics.PutTopicMetricCollectorAsync(collector, cancellationToken);

            builder = Diffusion.NewTopicMetricCollectorBuilder();
            builder = (ITopicMetricCollectorBuilder)builder.ExportsToPrometheus(true);
            builder = builder.GroupByTopicType(false);
            builder = builder.GroupByTopicView(true);
            builder = builder.GroupByPathPrefixParts(15);
            builder = (ITopicMetricCollectorBuilder)builder.MaximumGroups(250);
            collector = builder.Create("Topic Metric Collector 2", topicSelector);

            await session.Metrics.PutTopicMetricCollectorAsync(collector, cancellationToken);

            var listTopicMetricCollectors = await session.Metrics.ListTopicMetricCollectorsAsync(cancellationToken);

            foreach (var topicMetricCollector in listTopicMetricCollectors)
            {
                WriteLine($"{topicMetricCollector.Name}: " +
                    $"{topicMetricCollector.TopicSelector} " +
                    $"({topicMetricCollector.MaximumGroups}, " +
                    $"{GetAnswer(topicMetricCollector.ExportsToPrometheus)}, " +
                    $"{GetAnswer(topicMetricCollector.GroupsByTopicType)}, " +
                    $"{GetAnswer(topicMetricCollector.GroupsByTopicView)}, " +
                    $"{topicMetricCollector.GroupByPathPrefixParts})");
            }

            session.Close();
        }

        private string GetAnswer(bool result) => result ? "Yes" : "No";
    }
}

File: python/examples/src/diffusion_examples/server_configuration/metrics/topic_metric_collector/list_topic_metric_collectors.py
"""
Copyright © 2024 Diffusion Data Ltd.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import typing
import asyncio

from diffusion import sessions, Credentials
from diffusion.features.control.metrics.topic_metrics import (
    TopicMetricCollectorBuilder,
)

from diffusion_examples.utils.program import Example



class ListTopicMetricCollectors(Example):
    async def run(
        self,
        server_url: str = "<url>",
        principal: str = "<principal>",
        password: str = "<password>",
    ):
        async with sessions().principal(principal).credentials(
            Credentials(password)
        ).open(server_url) as session:
            topic_selector = "?my/topic//"

            # Create first collector
            builder = TopicMetricCollectorBuilder()
            builder.export_to_prometheus(False)
            builder.group_by_topic_type(True)
            builder.group_by_topic_view(True)
            builder.group_by_path_prefix_parts(15)
            builder.maximum_groups(10)
            collector = builder.create("Topic Metric Collector 1", topic_selector)
            await session.metrics.put_topic_metric_collector(collector)

            # Create second collector
            builder = TopicMetricCollectorBuilder()
            builder.export_to_prometheus(True)
            builder.group_by_topic_type(False)
            builder.group_by_topic_view(True)
            builder.group_by_path_prefix_parts(15)
            builder.maximum_groups(250)
            collector = builder.create("Topic Metric Collector 2", topic_selector)
            await session.metrics.put_topic_metric_collector(collector)
            list_topic_metric_collectors = (
                await session.metrics.list_topic_metric_collectors()
            )
            for topic_metric_collector in list_topic_metric_collectors:
                result_str = (
                    f"{topic_metric_collector.name}: "
                    f"{topic_metric_collector.topic_selector} "
                    f"({topic_metric_collector.maximum_groups}, "
                    f"{get_answer(topic_metric_collector.exports_to_prometheus)}, "
                    f"{get_answer(topic_metric_collector.groups_by_topic_type)}, "
                    f"{get_answer(topic_metric_collector.groups_by_topic_view)}, "
                    f"{topic_metric_collector.group_by_path_prefix_parts})"
                )
                print(result_str)


# noinspection PyMethodMayBeStatic
def get_answer(result: typing.Optional[bool]):
    return "Yes" if result else "No"


if __name__ == "__main__":
    asyncio.run(
        ListTopicMetricCollectors().run(
            server_url="ws://localhost:8080",
            principal="admin",
            password="password",
        )
    )
